runtime/jcl/common/java_lang_Class.cpp

```diff
@@ -1785,5 +1785,66 @@ Java_java_lang_Class_getNestHostImpl(JNIEnv *env, jobject recv)
 	vmFuncs->internalReleaseVMAccess(currentThread);
 	return result;
 }
+
+jobject JNICALL
+Java_java_lang_Class_getNestMembersImpl(JNIEnv *env, jobject recv)
+{
+	J9VMThread *currentThread = (J9VMThread*)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9MemoryManagerFunctions *mmFuncs = vm->memoryManagerFunctions;
+
+	j9object_t resultObject = NULL;
+	jobject result = NULL;
+
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+
+	J9Class *clazz = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, J9_JNI_UNWRAP_REFERENCE(recv));
+	J9Class *nestHost = clazz->nestHost;
+	J9ROMClass *romHostClass = nestHost->romClass;
+	U_16 nestMemberCount = romHostClass->nestMemberCount;
+
+	/*  Grab the Class<?> class for result object size */
+	J9Class *jlClass = J9VMJAVALANGCLASS(vm);
+	J9Class *arrayClass = fetchArrayClass(currentThread, jlClass);
+	if (NULL != currentThread->currentException) {
+		goto _throwException;
+	}
+
+	resultObject = mmFuncs->J9AllocateIndexableObject(currentThread, arrayClass, 1 + nestMemberCount, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+	if (NULL == resultObject) {
+		vmFuncs->setHeapOutOfMemoryError(currentThread);
+		goto _throwException;
+	}
+
+	/* Host class is always in zeroeth index */
+	J9JAVAARRAYOFOBJECT_STORE(currentThread, resultObject, 0, J9VM_J9CLASS_TO_HEAPCLASS(nestHost));
+
+	/* If host class claims nest members, they should be placed in second index onwards */
+	if (0 != nestMemberCount) {
+		J9SRP *nestMembers = J9ROMCLASS_NESTMEMBERS(romHostClass);
+		U_16 i = 0;
+		/* Classes in nest are in same runtime package & therefore have same classloader */
+		J9ClassLoader *classLoader = clazz->classLoader;
+
+		for (i = 0; i < nestMemberCount; i++) {
+			J9Class *nestMember = NULL;
+			J9UTF8 *nestMemberName = NNSRP_GET(nestMembers[i],J9UTF8*);
+
+			PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, resultObject);
+			nestMember = vmFuncs->internalFindClassUTF8(currentThread, J9UTF8_DATA(nestMemberName), J9UTF8_LENGTH(nestMemberName), classLoader, J9_FINDCLASS_FLAG_THROW_ON_FAIL);
+			resultObject = POP_OBJECT_IN_SPECIAL_FRAME(currentThread);
+			if (NULL != nestMember) {
```

Comment:
From the spec it seems like If you cant find a nestMember is a LinkageError expected. I think `internalFindClassUTF8` may set it to `ClassNotFoundException` which may not be right in this case.  

The spec also mentions that each nestMember must be validated to ensure that it has the correct nestHost.

Also, once an error is found, it might be better to stop at that point, rather then to continue looping. We should probably test this out with the OpenJDK build to see how they handle failures in this case.

Reply:
Thanks!

It looks like I could use `vmfuncs->vmFuncs->setCurrentExceptionUTF(currentThread, <nls message code> , <parameters to the nls message code>);`

The only examples of `setCurrentExceptionUTF` and a `LinkageError` that I'm seeing in the codebase all use `J9VMCONSTANTPOOL_JAVALANGLINKAGEERROR` - which I'm assuming should be related to constant pool issues. However some use cases of it (eg. `runtime/jcl/common/java_dyn_methodhandle.c/Java_java_lang_invoke_PrimitiveHandle_lookupField`, `runtime/vm/resolvesupport.cpp/resolveVirtualMethodRefInto`) don't look like they are dealing with constant pool issues.

Would I add a new NLS message into `runtime/nls/j9vm/j9vm.nls`, use the existing `J9VMCONSTANTPOOL_JAVALANGLINKAGEERROR`, or keep looking?

--

Also - I'm wondering what happens if a nest member doesn't claim the nest host, in this case. It should load+link without issue but should it raise a LinkageError or not? Adding a check for this could go in the same place - leaving a note here so I remember to check